File: LeetCode/Stack/java/simplifyPath.java
package LeetCode.Stack.java;

import java.util.Stack;

record simplifyPathRecord(String path, String expected) {
}

public class simplifyPath {
    public String simplifyPathSolution(String path) {
        Stack<String> stack = new Stack<>();
        StringBuilder sb = new StringBuilder("");

        for (char c : (path + "/").toCharArray()) {

            if (c == '/') {
                if (sb.toString().equals("..")) {

                    if (!stack.isEmpty()) {
                        stack.pop();
                    }
                } else if (!sb.toString().equals("")
                        && !sb.toString().equals(".")) {

                    stack.push(sb.toString());
                }
                sb = new StringBuilder("");
            } else {
                sb.append(String.valueOf(c));
            }
        }

        return "/" + String.join("/", stack);
    }

    public static void main(String[] args) {
        simplifyPathRecord input = new simplifyPathRecord("/home/", "/home");
        simplifyPath.testSolution(input);

        input = new simplifyPathRecord("/home//foo/", "/home/foo");
        simplifyPath.testSolution(input);

        input = new simplifyPathRecord("/home/user/Documents/../Pictures", "/home/user/Pictures");
        simplifyPath.testSolution(input);

        input = new simplifyPathRecord("/../", "/");
        simplifyPath.testSolution(input);

        input = new simplifyPathRecord("/.../a/../b/c/../d/./", "/.../b/d");
        simplifyPath.testSolution(input);
    }

    private static void testSolution(simplifyPathRecord input) {
        System.out.println("Input: path: " + input.path());
        System.out.println("Expected: " + input.expected());
        String res = new simplifyPath().simplifyPathSolution(input.path());
        System.out.println("Result: " + res);
        System.out.println(res.equals(input.expected()) ? "PASS" : "FAIL");
        System.out.println("-".repeat(50));
    }
}

File: LeetCode/Stack/python/simplifyPath.py
from typing import Dict


class Solution:
    def simplifyPath(self, path: str) -> str:
        stack = []
        cur = ""

        for c in path + "/":
            if c == "/":
                if cur == "..":
                    (len(stack) > 0) and stack.pop()
                elif cur != "" and cur != ".":
                    stack.append(cur)
                
                cur = ""
            else:
                cur += c

        return "/" + "/".join(stack)
    
    @staticmethod
    def testSolution(input: Dict[str, str]) -> None:
        print("Input: path: " + input["path"])
        print("Expected: " + str(input["expected"]))
        res = Solution().simplifyPath(input["path"])
        print("Result: " + str(res))
        print("PASS" if res == input["expected"] else "FAIL")
        print("-" * 50)
    

if __name__ == "__main__":
    input = {"path": "/home/", "expected": "/home"}
    Solution.testSolution(input)

    input = {"path": "/home//foo/", "expected": "/home/foo"}
    Solution.testSolution(input)

    input = {"path": "/home/user/Documents/../Pictures", "expected": "/home/user/Pictures"}
    Solution.testSolution(input)
    
    input = {"path": "/../", "expected": "/"}
    Solution.testSolution(input)

    input = {"path": "/.../a/../b/c/../d/./", "expected": "/.../b/d"}
    Solution.testSolution(input)

